### PR TITLE
Make [ENTER] default to submit the form

### LINKED ISSUE #82: [ENTER] key not intuitive
The ENTER key effectively clicks the nearest button, rather than submitting the form.  The nearest button may be "add" or "remove" depending if you're on the last item in a list.

For long forms, and/or people who prefer (or require) the keyboard, submitting the form is a bit annoying, and may result in accidental data deletion when hitting ENTER on a field.  The default behavior of an HTML form is that hitting ENTER within most fields will submit the form.

Possibly relevant: http://stackoverflow.com/questions/8294465/making-enter-key-on-an-html-form-submit-instead-of-activating-button

### PR DESCRIPTION
**Pull request by jechols**
Closes #82

### CODE REVIEW

File: app/assets/javascripts/hydra-editor/manage_repeating_fields.js [hunk 1]
```diff
@@ -8,8 +8,8 @@ var HydraEditor = (function($) {
           this.options = options;
 
           this.controls = $("<span class=\"input-group-btn field-controls\">");
-          this.remover  = $("<button class=\"btn btn-danger remove\"><i class=\"icon-white glyphicon-minus\"></i><span>Remove</span></button>");
-          this.adder    = $("<button class=\"btn btn-success add\"><i class=\"icon-white glyphicon-plus\"></i><span>Add</span></button>");
+          this.remover  = $("<button type=\"button\" class=\"btn btn-danger remove\"><i class=\"icon-white glyphicon-minus\"></i><span>Remove</span></button>");
```
Comment (jcoyne) [inline app/assets/javascripts/hydra-editor/manage_repeating_fields.js]: Why would you need to give a `button` a `type` of `button`?

**Comment by jcoyne** [comment]
This seems like a rather drastic change. I'd like to have someone from the Sufia camp review this. 
/cc @cam156 @mjgiarlo @awead @hectorcorrea

**Comment by awead** [comment]
I think this should follow what the useability standards are, which I would think would be to submit the form. @mtribone knows a bit more about this. Maybe he and weigh in too.

**Comment by hectorcorrea** [comment]
I've been puzzled myself on the current behavior of our forms when hitting ENTER. I like the idea of using the default browser behavior (submit) but definitively would like to hear @mtribone's thoughts.

File: app/assets/javascripts/hydra-editor/keyboard_support.js [hunk 2]
```diff
@@ -1,14 +0,0 @@
...
-        $addControl = $activeFieldControls.children('.add'),
-        $removeControl = $activeFieldControls.children('.remove');
-    if (event.keyCode == 13) {
-      event.preventDefault();
-      $addControl.click()
-      $removeControl.click()
-    }
-  });
-});
-
```
Comment (mtribone) [inline app/assets/javascripts/hydra-editor/keyboard_support.js]: Yeah our forms need a good looking into for usability/accessibility. So this code removes the default ability to add and remove metadata by hitting the enter key and gives it only to the button that submits the form. Does this mean that I have to use my mouse throughout the form to add and edit metadata or select items from a dropdown? And then get to use the enter key only to submit the form? If I'm following along, I think that there might be a better way to approach submitting the form using only a keyboard.

File: app/assets/javascripts/hydra-editor/manage_repeating_fields.js (hunk 1 above)
Comment (mtribone) [inline app/assets/javascripts/hydra-editor/manage_repeating_fields.js]: Type is a valid attribute of <button> in HTML 5 but "The button has no default behavior." If scripting is the only way this button is useful, we *could* be preventing people who use assistive devices from using the form.

File: app/assets/javascripts/hydra-editor/keyboard_support.js (hunk 2 above)
Comment (mtribone) [inline app/assets/javascripts/hydra-editor/keyboard_support.js]: It looks like the spacebar also works to activate controls. So if we bind the enter key to submit the form, this might be alright. I'm going to ping our local accessibility people and see how they navigate forms and activate controls.

File: app/assets/javascripts/hydra-editor/manage_repeating_fields.js (hunk 1 above)
Comment (jechols) [inline app/assets/javascripts/hydra-editor/manage_repeating_fields.js]: I'm not sure about the why - that just happened to be the recommended fix for browsers to know what to do on hitting [ENTER] when a form had lots of buttons.

File: app/assets/javascripts/hydra-editor/keyboard_support.js (hunk 2 above)
Comment (jechols) [inline app/assets/javascripts/hydra-editor/keyboard_support.js]: FYI, I have been tabbing to the button and hitting enter, as opposed to using the mouse.  The button is always one tab away from the input field, so it's pretty easy to get to.  The submit button was the tough one to navigate to since our forms are currently very big (and most fields are optional).

I'm not a keyboard-only user, mind you, I just happen to avoid the mouse as much as I can due to wrist issues.  Leaving the hands in one spot is just easier.

File: app/assets/javascripts/hydra-editor/manage_repeating_fields.js (hunk 1 above)
Comment (jcoyne) [inline app/assets/javascripts/hydra-editor/manage_repeating_fields.js]: Also on the page @mtribone mentioned:
> The type of the button. Possible values are:  submit: The button submits the form data to the server. This is the default if the attribute is not specified, or if the attribute is dynamically changed to an empty or invalid value.

So we probably should set it to `type=button` :+1:

Comment (jechols) [inline app/assets/javascripts/hydra-editor/manage_repeating_fields.js]: Ah, haha, I should have read that :)

File: app/assets/javascripts/hydra-editor/keyboard_support.js (hunk 2 above)
Comment (mtribone) [inline app/assets/javascripts/hydra-editor/keyboard_support.js]: Agreed. If you can keep your hands on the keys, why wouldn't you want to keep them there. And holy crow our forms are long! We might be able to alleviate that a little bit by adding tabindex values to jump over optional fields, but that might create additional additional problems. Or just redesign the forms. :-)

Let me double check on the spacebar/enter key usage and then I think we're good to go.

**Comment by mjgiarlo** [comment]
:+1:

File: app/assets/javascripts/hydra-editor/keyboard_support.js (hunk 2 above)
Comment (mtribone) [inline app/assets/javascripts/hydra-editor/keyboard_support.js]: I'm getting mixed results on the enter key and spacebar issue. I'd say "go for it" since it possibly alleviates accidentally deleting metadata depending on where the focus happens to be in the form.

Comment (jechols) [inline app/assets/javascripts/hydra-editor/keyboard_support.js]: Are you getting mixed responses from keyboard-only users?  Or are you saying you can't get tab+enter to properly "click" a button?

The former is a big issue and probably merits more discussion.  The latter, on the other hand, is something that's probably fixable.

Comment (mtribone) [inline app/assets/javascripts/hydra-editor/keyboard_support.js]: It's the former. Asking in a Yammer group at Penn State who designs or tests for accessibility. I'm looking for a few more people to weigh in and then I'm going to hit up some people individually.

Comment (jechols) [inline app/assets/javascripts/hydra-editor/keyboard_support.js]: I'm wondering what the preferred way would be to submit the form if ENTER were to default to the buttons.  I'd think they would want some kind of option to submit in a single keystroke, but maybe I'm just nuts here.  Or maybe there are assistive tools that render it a non-issue...?  As a mostly-but-not-100%-keyboard user, I can only comment on people like me who *prefer* the keyboard, but aren't forced to use it exclusively.

Comment (mtribone) [inline app/assets/javascripts/hydra-editor/keyboard_support.js]: They have to tab through everything to focus on it, so they can hear what it is before they use it. However, we can create access keys to get people directly to the submit button rather than tabbing through the whole thing. We also need to inform people ahead of time that we have some controls mapped to certain things

http://webaim.org/techniques/keyboard/accesskey#spec

So if we bind enter to the submit button, we should instruct folks at the top of the form to use the spacebar for the add/remove buttons, a defined access key, and enter to submit the form.

Comment (jechols) [inline app/assets/javascripts/hydra-editor/keyboard_support.js]: Okay, so it sounds like there's more accessibility work to be done.  Did you want me to put that stuff together in this PR?  Or can this be merged and a new PR be made for addressing the other issues?

It *sounds* like everybody agrees that "enter submits form" is better than "enter clicks nearest button", at least.

Comment (mtribone) [inline app/assets/javascripts/hydra-editor/keyboard_support.js]: I'd say let's put the accessibility work into another PR.

Comment (jechols) [inline app/assets/javascripts/hydra-editor/keyboard_support.js]: I'll start on another PR for the rest of the accessibility stuff.  @mjgiarlo, @jcoyne - is this reasonable to merge?

**Comment by mjgiarlo** [comment]
Any further thoughts, @jcoyne @mtribone @awead @hectorcorrea?  I'm :+1: to merging, methinks.

**Comment by mtribone** [comment]
@mjgiarlo I would have already pushed the button if I had wizard class permissions.
